shell/common/rasterizer.cc

```diff
@@ -214,16 +214,15 @@ RasterStatus Rasterizer::Draw(
   // between successive tries.
   switch (consume_result) {
     case PipelineConsumeResult::MoreAvailable: {
-      delegate_.GetTaskRunners().GetRasterTaskRunner()->PostTask(
-          fml::MakeCopyable(
-              [weak_this = weak_factory_.GetWeakPtr(), pipeline,
-               resubmit_recorder = std::move(resubmit_recorder),
-               discard_callback = std::move(discard_callback)]() mutable {
-                if (weak_this) {
-                  weak_this->Draw(std::move(resubmit_recorder), pipeline,
-                                  std::move(discard_callback));
-                }
-              }));
+      if (raster_status != RasterStatus::kSuccess)
```

Comment:
Because Rasterizer::Draw post task with old timestamp, i guess the purpose of this behavior isn't general and a  constraint should be added there to prevent old timestamps match a fresh pipeline item.

Reply:
@yeran1212 
I don't think this change is correct. We should call the `Rasterizer::Draw` method every time when `PipelineConsumeResult::MoreAvailable`. The situation you encounter is that the frame is rasterized successfully, but the `LayerTreePipeline` is not empty, so need to rasterize the next frame, but at this time it should actually pass the `frame_timings_recorder` related the next frame as param instead of `resubmit_recorder`.

Unfortunately, in the current implementation, `LayerTreePipeline` only includes layer tree, and does not include `frame_timings_recorder` or timestamps which can used by making `frame_timings_recorder`. 

It seems that we should save the `frame_timings_recorder` created in the method `Animator::Render` so that the Rasterizer can find it in that case.

Reply:
@ColdPaleLight 
I agree with you that We should call the `Rasterizer::Draw` method every time when `PipelineConsumeResult::MoreAvailable`. But the situation is frame is rasterized successfully, but the LayerTreePipeline is not empty meanwhile the remained pipeline item have already had corresponding taskA in `RasterizerTaskRunner` due to `Animator::Render`.  the corresponding taskA will invoke `Rasterizer::Draw`. should we post a taskB again even if we save the frame_timings_recorder created in the method Animator::Render.  i think taskB is redundant.

Reply:
Maybe you can introduce a new variable to determine if `Rasterizer::Draw` needs to be called instead of according the result of expression `consume_result
== PipelineConsumeResult::MoreAvailable && raster_status != RasterStatus::kSuccess`. Otherwise it's really confusing. I'm not sure if this is a good solution though. Maybe @chinmaygarde 
 and @iskakaushik  can enlighten us.

Reply:
@ColdPaleLight 
According to code below, only when raster_status is `kResubmit`, `kSkipAndRetry`, `kEnqueuePipeline` `Rasterizer::Draw` need post task to `RasterizerTaskRunner`.
https://github.com/flutter/engine/blob/7b5edec4cd13fd948fa12c0cacdf5d1852b0149f/shell/common/rasterizer.cc#L191-L205
https://github.com/flutter/engine/blob/7b5edec4cd13fd948fa12c0cacdf5d1852b0149f/shell/common/rasterizer.cc#L236-L239
when raster_status is `kResubmit`, `kSkipAndRetry` resubmitted_layer_tree_ will be added to pipeline, `resubmitted_layer_tree_` with `resubmit_recorder` is right

when raster_status is `kEnqueuePipeline` one more taskB is needed to make sure item(B) is consumed at right thread. because of previous taskB is yielded.
https://github.com/flutter/engine/blob/7b5edec4cd13fd948fa12c0cacdf5d1852b0149f/shell/common/rasterizer.cc#L459-L480

Therefore, i added `raster_status != RasterStatus::kSuccess`.

@iskakaushik
But i am curious is there any possible that animator produce (C) to the pipeline and add taskC into `RasterizerTaskRunner` between old taskB is consumed and new taskB is resubmitted.  if that item B will match taskC and item C will match taskB.